pkg/node/local_node_store.go

```diff
@@ -72,113 +46,75 @@ type LocalNodeStoreParams struct {
-	// that synchronously calls into all subscribers, which is a potentially
-	// expensive operation, and a possible source of deadlocks (e.g., one of
-	// the subscribers needs to acquire another mutex, which is held by a
-	// separate goroutine trying to call LocalNodeStore.Get()). In addition,
-	// getMu also guards the complete field, as it is used by Get() to
-	// determine that the LocalNodeStore was stopped. When both mu and getMu
-	// are to be acquired together, mu shall be always acquired first.
-	mu    lock.Mutex
-	getMu lock.RWMutex
-
-	value    LocalNode
-	hasValue <-chan struct{}
-	emit     func(LocalNode)
-	complete func(error)
+	db    *statedb.DB
+	nodes statedb.RWTable[*LocalNode]
 }
 
-func NewTestLocalNodeStore(mockNode LocalNode) *LocalNodeStore {
-	src, emit, complete := stream.Multicast[LocalNode](stream.EmitLatest)
-	emit(mockNode)
-	return &LocalNodeStore{
-		Observable: src,
-		emit:       emit,
-		complete:   complete,
-		value:      mockNode,
-		hasValue: func() <-chan struct{} {
-			ch := make(chan struct{})
-			close(ch)
-			return ch
-		}(),
+func NewLocalNodeStore(params LocalNodeStoreParams) (*LocalNodeStore, error) {
+
+	initNode := &LocalNode{
+		Node: types.Node{
+			// Explicitly initialize the labels and annotations maps, so that
+			// we don't need to always check for nil values.
+			Labels:      make(map[string]string),
+			Annotations: make(map[string]string),
+			Source:      source.Unspec,
+		},
+		Local: &LocalNodeInfo{},
 	}
-}
 
-func NewLocalNodeStore(params LocalNodeStoreParams) (*LocalNodeStore, error) {
-	src, emit, complete := stream.Multicast[LocalNode](stream.EmitLatest)
-	hasValue := make(chan struct{})
-
-	s := &LocalNodeStore{
-		Observable: src,
-		value: LocalNode{
-			Logger: params.Logger,
-			Node: types.Node{
-				// Explicitly initialize the labels and annotations maps, so that
-				// we don't need to always check for nil values.
-				Labels:      make(map[string]string),
-				Annotations: make(map[string]string),
-				Source:      source.Unspec,
-			}},
-		hasValue: hasValue,
+	var initDone func(statedb.WriteTxn)
+	wtxn := params.DB.WriteTxn(params.Nodes)
+	if params.Sync != nil {
+		// Register an initializer if a LocalNodeSynchronizer is given
+		initDone = params.Nodes.RegisterInitializer(wtxn, "LocalNodeSynchronizer")
+	} else {
+		// No synchronizer, insert the initial node immediately.
+		params.Nodes.Insert(wtxn, initNode)
 	}
+	wtxn.Commit()
 
-	bctx, cancel := context.WithCancel(context.Background())
-	var wg sync.WaitGroup
+	s := &LocalNodeStore{params.DB, params.Nodes}
 
 	params.Lifecycle.Append(cell.Hook{
 		OnStart: func(ctx cell.HookContext) error {
-			s.mu.Lock()
-			defer s.mu.Unlock()
 			if params.Sync != nil {
-				if err := params.Sync.InitLocalNode(ctx, &s.value); err != nil {
+				wtxn := params.DB.WriteTxn(params.Nodes)
+				err := params.Sync.InitLocalNode(ctx, initNode)
+				params.Nodes.Insert(wtxn, initNode)
```

Comment:
The behavior is different compared to how it was before, as in that case we had a valid entry also before the initialization completed (i.e., this start hook run), which could be potentially mutated via `Invoke` functions, and the initialization here respected the possible changes. In this case, instead, those changes would be blindly overwritten when performing the insertion. It may be worthy either preserving the previous behavior (i.e., inserting the node immediately, and then retrieving the current version of `InitLocalNode`) or double-checking that no code path is currently relying on that behavior.

Reply:
Yeah I was struggling a bit with this exact thing. I'll look at doing it so that I insert before the start and just make sure any getters or observers block on initialized.